Add vacuole upgrade to store double of only one compound

ender-titan1 opened the pull request:
**Brief Description of What This PR Does**

This PR modifies CompoundBag to enable it to handle separate capacities for each compound,  as well as adding an upgrade to the vacuole to enable it to have double the capacity but only be able to handle one type of compound.

**Related Issues**

Closes #4109 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

hhyyrylainen commented on src/microbe_stage/CompoundBag.cs:
```diff
@@ -13,20 +13,23 @@ public class CompoundBag : ICompoundStorage
     /// </summary>
     [JsonProperty]
-    public float Capacity { get; set; }
+    public Dictionary<Compound, float> CompoundCapacities { get; private set; } = new();
```
The internal details of how the capacity is stored should not be leaked. This should be made private. I created the `GetCapacityForCompound` method for a reason. And second problem is that if a mod for example adds more compounds or we add more compounds in a later update, this is so going to blow up and cause a game crash somewhere.

So here's my suggestion have a "nominal capacity" private field and a nullable dictionary of specific capacities also a private field. Then modify the GetCapacityForCompound to first check if a specific capacity exists for a compound, and if it does return it, otherwise return the nominal capacity. I think with those changes we'd have a design that I can think of being good enough.

ender-titan1 replied:
Ok, on it

ender-titan1 commented:
I've implemented the requested changes, however, these changes created a problem where microbes would spawn without any initial compounds. I have fixed this in a somewhat hacky way and still do not know what exactly was causing the problem. I also wonder if `organellesCapacity` is still required in `Microbe.Internal.cs` since the capacity field had been made private.

hhyyrylainen commented:
Regarding the bug with initial compounds, I've been converting that code over to a new system on my branch this week and I remember reading comments that the initial compound add used a direct add of compounds. That way it bypasses any normal limits or usefulness check. Could you modify your code to go back to that way of doing that to not need the workaround?

ender-titan1 commented:
So if I understand correctly, I should create a new method in `CompoundBag` to add compounds without checking capacity

hhyyrylainen commented:
Can your work wait until my branch is merged as I already wrote this:
```
    public void AddInitialCompounds(IReadOnlyDictionary<Compound, float> compounds)
    {
        foreach (var entry in compounds)
        {
            if (!Compounds.TryGetValue(entry.Key, out var existingAmount))
            {
                Compounds[entry.Key] = entry.Value;
                continue;
            }

            float toAdd = entry.Value - existingAmount;

            if (toAdd > 0)
                Compounds[entry.Key] = existingAmount + toAdd;
        }
    }
```

doing it differently twice will lead to nice headache merge conflicts for me...

If I remember right didn't the initial compounds just use the `Compounds.Compounds` dictionary directly to set the values? That just goes past any limits etc.

ender-titan1 commented:
Yes, I can wait until that branch has been merged

hhyyrylainen commented:
If wait until then you'll need to redo Microbe.Interior.cs, MicrobeAI.cs and ProcessSystem.cs changes (or solve very difficult merge conflicts as most of those files will be gone or moved).

I think something like:
```
// Temporary code before proper initial compounds set method is added to CompoundBag
Compounds.Compounds.TryGetValue(compound, out var existing);
Compounds.Compounds[compound] = existing + stuffToAdd;
```
Would be simple enough and this PR could move forward without having to wait a month, which is my maybe a realist estimate on when my work would be completed.

ender-titan1 commented:
I don't quite understand what you mean by "redo changes". Should I revert those files back to the original? And the code provided is meant to be inserted in the `SetInitialCompounds` method if I am to understand correctly.

hhyyrylainen commented:
I'm just warning that as I'm making major changes, when those changes are merged to master that will lead to a large number of merge conflicts in your PR. So completing the functionality of the PR now with that temporary change will be less work for you. And then of course cleaning up the changes a bit to get this ready for merge, as I noticed at least a few places I'm going to comment on the design approach once this is functionally ready.

But if you are fine with that, we can get back to this PR in a month and then get this ready by fixing all the new merge conflicts with master.

hhyyrylainen commented on src/microbe_stage/organelle_components/StorageComponent.cs:
```diff
@@ -29,3 +29,22 @@ public void Check(string name)
+        IsSpecialized = isSpecialized;
+    }
+
+    public Compound Specialization { get; set; }
```
Is there a reason for using two variables? I'd just make one property `Compound? SpecializedFor` and make it so that null means no specialization and then a valid compound value is specialization for that compound.

hhyyrylainen commented on src/late_multicellular_stage/MulticellularCreature.cs:
```diff
@@ -489,84 +489,85 @@ public bool AttemptInteraction(IInteractableEntity target, InteractionType inter
-                        $"Only player creature can open crafting ({nameof(RequestCraftingInterfaceFor)} is unset)");
-                    return false;
-                }
+                    if (RequestCraftingInterfaceFor == null)
```
Looks like the indentation got messed up in this file.

hhyyrylainen commented on src/microbe_stage/CompoundBag.cs:
```diff
@@ -12,21 +12,16 @@ public class CompoundBag : ICompoundStorage
-        Capacity = capacity;
-    }
+    [JsonProperty]
+    private float nominalCapacity;
```
I don't see why this would need to be a field. Can't this be a property? That avoid having a "useless" manually written setter for the nominal capacity.

hhyyrylainen commented on src/microbe_stage/CompoundBag.cs:
```diff
@@ -38,13 +33,30 @@ public CompoundBag(float capacity)
+    }
+
+    /// <summary>
+    ///   Sets the capacity for all compounds
```
I think this comment needs to elaborate that this is the capacity for all compounds that don't have a specific capacity set.

hhyyrylainen commented on src/microbe_stage/Microbe.Interior.cs:
```diff
@@ -1,6 +1,7 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using AngleSharp.Dom;
```
I'm pretty sure this is a mistake, the microbe code should not be doing any HTML parsing...

hhyyrylainen commented:
I think you hit a bunch of warnings that were hiding underneath the surface on master some time ago. I think those should go away if you merge master into your branch.

hhyyrylainen commented on src/microbe_stage/organelle_components/StorageComponent.cs:
```diff
@@ -29,3 +29,19 @@ public void Check(string name)
+        SpecializationFor = specializationFor;
+    }
+
+    public Compound? SpecializationFor { get; set; }
```
This doesn't sound grammatically correct to me, is it? To me `SpecializedFor` would sound grammatically correct when reading code aloud that uses this property. Would that be better?

hhyyrylainen commented on src/microbe_stage/CompoundBag.cs:
```diff
@@ -12,21 +12,20 @@ public class CompoundBag : ICompoundStorage
     /// <summary>
-    ///   How much of each compound this bag can store.
-    ///   Currently a CompoundBag can hold the same amount of each compound.
+    ///   Stores the default capacity for all compounds that do
```
```suggestion
    ///   Specifies the default capacity for all compounds that do
```
As this variable doesn't "store" the data just specifies how much something can be stored elsewhere.

hhyyrylainen commented on src/microbe_stage/CompoundBag.cs:
```diff
@@ -38,13 +37,25 @@ public CompoundBag(float capacity)
     /// </summary>
-    /// <returns>Returns <see cref="Capacity"/> if the compound is useful, otherwise 0</returns>
+    /// <returns>
+    ///   Returns compoundCapacities[compound] if the compound has a specific capacity set,
+    ///   or nominalCapacity if the compound is useful, otherwise 0
```
```suggestion
    ///   Returns the capacity this bag has for storing the compound if it is useful, otherwise 0
```
I'd omit the internal details, especially the non `<see ...` type reference to a private variable.

hhyyrylainen commented on src/microbe_stage/Microbe.Interior.cs:
```diff
@@ -1323,16 +1332,58 @@ private void OnOrganelleRemoved(PlacedOrganelle organelle)
+    }
+
+    /// <summary>
+    ///   Updates <see cref="organellesCapacity"/>
```
I'm noticing that you are wrapping the documentation text really tightly? Is there a reason? The documentation comments should be filled all the way to column 120.

hhyyrylainen commented on src/microbe_stage/Microbe.Interior.cs:
```diff
@@ -1323,16 +1332,58 @@ private void OnOrganelleRemoved(PlacedOrganelle organelle)
+        Compounds.NominalCapacity = organellesCapacity;
+
+        foreach (var entry in additionalCompoundCapacities)
+            Compounds.SetCapacityForCompound(entry.Key, entry.Value + organellesCapacity);
```
This doesn't remove special capacities that are now gone from the cell (for example if the player switches a vacuole from glucose to ammonia, they get the extra capacity for both as the old value is never seemingly removed). There needs to be a method in the compound bag to reset the special capacities before they are re-applied to ensure no outdated values exist.

ender-titan1 replied:
Regarding this, the feature still works as intended. This is because `additionalCompoundCapacities` for in your example glucose is set to zero when the capacity is begin calculated. This means that while the entry exists in the dictionary, the value is zero. I have also play-tested this and the feature works as intended.

hhyyrylainen replied:
Ah, you mean with `UpdateCapacity`? I guess that is true, though with the way I'm refactoring the microbe stage the "organelle removed" callback will be entirely removed, so could you futureproof your code by doing that my suggestion? That way I don't accidentally introduce this bug when trying to fix merge conflicts with my branch.

hhyyrylainen commented on src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:
```diff
@@ -0,0 +1,141 @@
+
+    public void OnStartFor(OrganelleTemplate organelle, GameProperties currentGame)
+    {
+        shownChoices = SimulationParameters.Instance.GetAllCompounds().Values.ToList();
```
I don't see you modifying the shown choices so you should be able to just use the returned IReadOnlyList directly here instead of taking a copy of the data.

ender-titan1 replied:
Forgive me but I do not understand, `GetAllCompounds` returns a dictionary and the `Values` property returns a ValueCollection that I cannot convert to a list. I am a bit confused and ask for clarification.

hhyyrylainen replied:
My mistake, I thought GetAllCompounds would return a `IReadOnlyList<Compound>`. Alright then this is a fine approach without any changes.

hhyyrylainen commented on src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:
```diff
@@ -0,0 +1,141 @@
+
+        // Select glucose by default
+        var defaultCompoundIndex =
+            shownChoices.FindIndex(c => c.InternalName == "glucose");
```
Don't the other upgrade GUIs have the default value stored in Constants? This should also be in constants and named something like `DEFAULT_VACUOLE_UPGRADE_FOR_COMPOUND = "glucose"`.

hhyyrylainen commented on src/microbe_stage/editor/upgrades/VacuoleUpgradeGUI.cs:
```diff
@@ -0,0 +1,141 @@
+
+            compounds.Selected = specialization != null ?
+                shownChoices.IndexOf(specialization) :
+                shownChoices.FindIndex(c => c.InternalName == "glucose");
```
Shouldn't this be `defaultCompoundIndex`? to not have to find the same index again.

hhyyrylainen commented:
A change was just merged that requires adding a comparison method for the upgrade data: https://github.com/Revolutionary-Games/Thrive/pull/4463 so I suspect that that'll need a small tweak to this PR to make sure this compiles (with latest master).

hhyyrylainen commented on RevolutionaryGamesCommon:
Does this PR accidentally downgrade the common module version? It seems like this is moving it back to version `f1b5b669dbce77e203bc001bc662d2d549b67a40` even though it was newer version before on the master branch.

hhyyrylainen commented on src/microbe_stage/Microbe.Interior.cs:
```diff
@@ -1323,16 +1331,55 @@ private void OnOrganelleRemoved(PlacedOrganelle organelle)
+    ///   to take into account the addition or removal of an organelle
+    /// </summary>
+    /// <param name="organelle">The organelle being placed or removed</param>
+    /// <param name="negative"> Should be true if the organelle is being removed, otherwise false </param>
```
```suggestion
    /// <param name="negative">Should be true if the organelle is being removed, otherwise false</param>
```

hhyyrylainen commented:
Just those two minor things I still noticed (well one is possibly not minor as downgrading the common submodule is a bit of real problem). Once those are fixed I think I'll just merge this, unless I can manage to get someone to playtest these changes.

buckly90 commented:
Seems to work perfectly well, with one unfortunate exception.

Iron storage is not being increased as it should be, and the upgrade has no appearant effect.

ender-titan1 commented:
Acknowledged, I will see what is causing the iron capacity not to increase, and I will try to resolve the submodule problem as soon as i can.

ender-titan1 commented:
I've fixed the iron issue, however, I do not know how to fix the submodule problem. I've never worked with submodules in the past and trying to update them using git bash creates an error in the code, so I really don't know what I am supposed to do

hhyyrylainen commented:
> I've never worked with submodules in the past and trying to update them using git bash creates an error in the code, so I really don't know what I am supposed to do

Well that explains it. You've pulled in new code and then committed *all changes* which incorrectly picked the old submodule version. Git pull or branch switching do not update submodules, you need to run `git submodule update` to update submodules. You can probably fix things by:
```
cd RevolutionaryGamesCommon
git checkout 3a737999e08a88cd728067a7f4be695ef1d65f10
cd ..
git add ...
git commit ...
```

to create a new commit that sets the submodule version to what it is on master.

hhyyrylainen reviewed (approved):
Code changes look good now. Also the fix to the iron probably fixed all vacuole configuration issues from the underlying problem of the Godot option button not having the same items as the internal compounds list.

hhyyrylainen commented:
@buckly90 do you want to re-test this to make sure the iron bug is fixed? That's the last potential thing to test before merging this. At the latest I'll merge this in two days to make sure this makes it into the BOTD.

buckly90 commented:
Looks great, splendid job!
